src/components/implementation/no_interface/llbooter/boot_deps.h

```diff
@@ -94,73 +106,242 @@ boot_newcomp_sinv_alloc(spdid_t spdid)
+	}
+
+	if (is_sched) {
+		sl_thd_param_set(thd, sched_param_pack(SCHEDP_BUDGET, THD_BUDGET * 1000));
+		sl_thd_param_set(thd, sched_param_pack(SCHEDP_WINDOW, THD_PERIOD * 1000));
+	}
 }
 
 static void
 boot_bootcomp_init(void)
 {
-	cos_meminfo_init(&boot_info.mi, BOOT_MEM_KM_BASE, COS_MEM_KERN_PA_SZ, BOOT_CAPTBL_SELF_UNTYPED_PT);
+	cos_defcompinfo_init();
+	boot_info = cos_compinfo_get(cos_defcompinfo_curr_get());
+	assert(boot_info);
 
-	cos_compinfo_init(&boot_info, BOOT_CAPTBL_SELF_PT, BOOT_CAPTBL_SELF_CT, BOOT_CAPTBL_SELF_COMP,
-	                  (vaddr_t)cos_get_heap_ptr(), BOOT_CAPTBL_FREE, &boot_info);
+	cos_meminfo_init(&(boot_info->mi), BOOT_MEM_KM_BASE, COS_MEM_KERN_PA_SZ,
+			BOOT_CAPTBL_SELF_UNTYPED_PT);
+	sl_init(SL_MIN_PERIOD_US);
 }
 
 static void
 boot_done(void)
 {
-	printc("Booter: done creating system.\n\n");
-	cos_thd_switch(schedule[sched_cur]);
+	printc("Booter: done creating system.\n");
+	printc("Booter: Starting SL event loop\n\n");
+	sl_sched_loop();
+}
+
+static int
+boot_check_scheduler(char *comp_name)
+{
+	int i;
+	char *prefix = "sl_";
+
+	for (i = 0; i < 3; i++) {
+		if (comp_name[i] == '\0' || comp_name[i] != prefix[i]) 
+			return 0;
+	}
+
+	return 1;
+}
+
+static int
+boot_check_shdmem(char *comp_name)
+{
+	int i;
+	char *prefix = "shmem";
+
+	for (i = 0; i < 5; i++) {
+		if (comp_name[i] == '\0' || comp_name[i] != prefix[i])
+			return 0;
+	}
+
+	return 1;
 }
 
-/* Run after a componenet is done init execution, via sinv() into booter */
 void
 boot_thd_done(void)
 {
-	sched_cur++;
+	printc("Done booting\n\n");
+	while(1) sl_thd_block(0);
+}
 
-	if (schedule[sched_cur] != 0) {
-		cos_thd_switch(schedule[sched_cur]);
-	} else {
-		printc("Done Initializing\n");
+void
+boot_pgtbl_cap_transfer(int dst, int src, int cap_slot)
+{
+	printc("booter transfering pgtbl cap: %lu to: %d, from: %d, into: %d...",
+			new_comp_cap_info[src].compinfo->pgtbl_cap, dst, src, cap_slot);
+
+	cos_cap_cpy_at(new_comp_cap_info[dst].compinfo, cap_slot, boot_info,
+			new_comp_cap_info[src].compinfo->pgtbl_cap);
+
+	printc("done\n");
+}
+
+void
+boot_comp_cap_transfer(int dst, int src, int cap_slot)
+{
+	printc("booter transfering comp cap: %lu to: %d, from: %d, into: %d...",
+			new_comp_cap_info[src].compinfo->comp_cap, dst, src, cap_slot);
+
+	cos_cap_cpy_at(new_comp_cap_info[dst].compinfo, cap_slot, boot_info,
+			new_comp_cap_info[src].compinfo->comp_cap);
+
+	printc("done\n");
+}
+
+
+void *
+boot_sinv_fn(boot_sinv_op op, void *arg1, void *arg2, void *arg3)
+{
+	void *ret = NULL;
+
+	switch (op) {
+		case INIT_DONE:
+			/*
+			 * DEPRICATED:
+			 * Should not be needed as llbooter is not scheduler
+			 * Kept for the purpose of older test code like llbooter_pong.sh
+			 */
+			boot_thd_done();
+			break;
+		case REQ_PGTBL_CAP:
+			/* arg1: dst comp, arg2: src comp, arg3: cap_slot */
+			boot_pgtbl_cap_transfer((int)arg1, (int)arg2, (int)arg3);
+			ret = (void *)0;
+			break;
+		case REQ_NUM_COMPS:
+			ret = (void *)num_cobj;
```

Comment:
Do we want to return the count *including* the scheduler?  Isn't the scheduler asking for how many *other* components there are?

Reply:
yes good point. This is also used by the memmgr when needing to allocate caps for pgtbls